File: ai-solutions/01/sonnet-4.5/js/solution.js
const fs = require('fs');

function countClicksOnZero(start, signedSteps) {
    if (signedSteps === 0) {
        return 0;
    }
    
    const startNorm = ((start % 100) + 100) % 100;
    
    if (signedSteps > 0) {
        const numClicks = signedSteps;
        const firstZeroClick = startNorm === 0 ? 100 : 100 - startNorm;
        
        if (firstZeroClick > numClicks) {
            return 0;
        }
        
        return 1 + Math.floor((numClicks - firstZeroClick) / 100);
    } else {
        const numClicks = -signedSteps;
        const firstZeroClick = startNorm === 0 ? 100 : startNorm;
        
        if (firstZeroClick > numClicks) {
            return 0;
        }
        
        return 1 + Math.floor((numClicks - firstZeroClick) / 100);
    }
}

function solve() {
    const input = fs.readFileSync('../../../../inputs/01.txt', 'utf8').trim();
    const lines = input.split('\n');
    
    let position = 50;
    let part1Count = 0;
    let part2Count = 0;
    
    for (const line of lines) {
        const direction = line[0];
        const distance = parseInt(line.slice(1));
        const signedSteps = direction === 'R' ? distance : -distance;
        
        // Count zeros passed during rotation (Part 2)
        part2Count += countClicksOnZero(position, signedSteps);
        
        // Update position
        position = (position + signedSteps) % 100;
        position = (position + 100) % 100;
        
        // Check if we ended at 0 (Part 1)
        if (position === 0) {
            part1Count++;
        }
    }
    
    console.log(part1Count);
    console.log(part2Count);
}

solve();

File: ai-solutions/01/sonnet-4.5/python/solution.py
#!/usr/bin/env python3

def count_clicks_on_zero(start: int, signed_steps: int) -> int:
    """
    Count how many *clicks* during this rotation cause the dial to point at 0.
    This includes the final click if the rotation ends on 0.
    If the dial starts at 0, we only count reaching 0 again after a full 100 clicks.
    """
    if signed_steps == 0:
        return 0

    start_norm = ((start % 100) + 100) % 100

    if signed_steps > 0:
        num_clicks = signed_steps
        first_zero_click = 100 if start_norm == 0 else 100 - start_norm
        if first_zero_click > num_clicks:
            return 0
        return 1 + (num_clicks - first_zero_click) // 100
    else:
        num_clicks = -signed_steps
        first_zero_click = 100 if start_norm == 0 else start_norm
        if first_zero_click > num_clicks:
            return 0
        return 1 + (num_clicks - first_zero_click) // 100

def solve():
    with open('../../../../inputs/01.txt', 'r') as f:
        lines = f.read().strip().split('\n')
    
    position = 50
    part1_count = 0
    part2_count = 0
    
    for line in lines:
        direction = line[0]
        distance = int(line[1:])

        signed_steps = distance if direction == 'R' else -distance

        part2_count += count_clicks_on_zero(position, signed_steps)
        position = (position + signed_steps) % 100
        position = (position + 100) % 100

        # Part 1: count times the dial is left at 0 after a rotation
        if position == 0:
            part1_count += 1
    
    print(part1_count)
    print(part2_count)

if __name__ == '__main__':
    solve()

